p2p/router.go

```diff
@@ -0,0 +1,557 @@
+//
+// FIXME: providing protocol/transport maps is cumbersome in tests, we should
+// consider adding Protocols() to the Transport interface instead and register
+// protocol/transport mappings automatically on a first-come basis.
+func NewRouter(logger log.Logger, transports map[Protocol]Transport, peers []PeerAddress) *Router {
+	router := &Router{
+		logger:          logger,
+		transports:      transports,
+		store:           newPeerStore(),
+		stopCh:          make(chan struct{}),
+		channelQueues:   map[ChannelID]queue{},
+		channelMessages: map[ChannelID]proto.Message{},
+		peerQueues:      map[NodeID]queue{},
+		peerUpdatesCh:   make(chan PeerUpdate, 1),
+		peerUpdatesSubs: map[*PeerUpdatesCh]*PeerUpdatesCh{},
+	}
+	router.BaseService = service.NewBaseService(logger, "router", router)
+
+	for _, address := range peers {
+		if err := router.store.Add(address); err != nil {
+			logger.Error("failed to add peer", "address", address, "err", err)
+		}
+	}
+
+	return router
+}
+
+// OpenChannel opens a new channel for the given message type. The caller must
+// close the channel when done, and this must happen before the router stops.
+func (r *Router) OpenChannel(id ChannelID, messageType proto.Message) (*Channel, error) {
+	// FIXME: NewChannel should take directional channels so we can pass
+	// queue.dequeue() instead of reaching inside for queue.queueCh.
+	queue := newFIFOQueue()
+	channel := NewChannel(id, messageType, queue.queueCh, make(chan Envelope), make(chan PeerError))
+
+	r.channelMtx.Lock()
+	defer r.channelMtx.Unlock()
+
+	if _, ok := r.channelQueues[id]; ok {
+		return nil, fmt.Errorf("channel %v already exist", id)
+	}
+	r.channelQueues[id] = queue
+	r.channelMessages[id] = messageType
+
+	go func() {
+		defer func() {
+			r.channelMtx.Lock()
+			delete(r.channelQueues, id)
+			delete(r.channelMessages, id)
+			r.channelMtx.Unlock()
+			queue.close()
+		}()
+		r.routeChannel(channel)
+	}()
+
+	return channel, nil
+}
+
+// routeChannel receives outbound messages and errors from a channel and routes
+// them to the appropriate peer. It returns when either the channel is closed or
+// the router is shutting down.
+func (r *Router) routeChannel(channel *Channel) {
+	for {
+		select {
+		case envelope, ok := <-channel.outCh:
+			if !ok {
+				return
+			}
+
+			// FIXME: This is a bit unergonomic, maybe it'd be better for Wrap()
+			// to return a wrapped copy.
+			if _, ok := channel.messageType.(Wrapper); ok {
+				wrapper := proto.Clone(channel.messageType)
+				if err := wrapper.(Wrapper).Wrap(envelope.Message); err != nil {
+					r.Logger.Error("failed to wrap message", "err", err)
+					continue
+				}
+				envelope.Message = wrapper
+			}
+			envelope.channel = channel.id
+
+			if envelope.Broadcast {
+				r.peerMtx.RLock()
+				peerQueues := make(map[NodeID]queue, len(r.peerQueues))
+				for peerID, peerQueue := range r.peerQueues {
+					peerQueues[peerID] = peerQueue
+				}
+				r.peerMtx.RUnlock()
```

Comment:
nit: this method is verbose - maybe create and move this to a `Router#copyPeerQueues`?

Reply:
I agree it's too verbose. But this code is going to change quite a bit as the refactor progresses, so I figured there's no point cleaning it up just yet since there may be structural changes anyway.